Feature/qfeed 52 validate question duplicate

jhk01007 opened the pull request:
<!--
  템플릿은 아직 PR 작성이 익숙하지 않으신 분들을 위해서 제공하는 가이드입니다!
  리뷰어 또는 이 PR을 보게 될 다른 사람들이 이 PR을 보는데 참고할 수 있는 내용이 있다면 포함해서 작성해주시면 됩니다.
-->
## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->
질문을 생성하고 질문의 중복을 검증하는 로직을 구현하였습니다.

## 👩‍💻 요구 사항과 구현 내용 <!-- 기능을 Commit 별로 잘개 쪼개고, Commit 별로 설명해주세요 -->
- Category @Getter 추가 및 정적 팩토리 메서드 추가
- SQL 쿼리 로그 설정 추가
- Redis Vector Database & Search를 통해 질문의 벡터값을 저장하고 검색(AI가  생성한 질문의 중복 검증을 위해)
- 23시 30분에 질문이 생성되도록 스케줄링
- 중복된 질문들(기존에 저장된 질문, AI가 이전에 생성한 질문)을 동적으로 프롬프트에 질문 블랙리스트로 추가하여 AI가 반복된 질문을 생성하지 않게 조정
- 모든 카테고리의 질문 생성이 완료되면 데이터베이스(Vector DB, RDB)에 일괄적으로 저장
![image](https://github.com/user-attachments/assets/505a10e2-8a93-4bbf-a209-c515ffac5f51)
- 사용된 토큰을 계산하는 유틸 추가
<img width="463" alt="image" src="https://github.com/user-attachments/assets/840bcc92-3df2-42e9-ac78-fe14f6ad1fc6">


## ✅ 피드백 반영사항  <!-- 지난 코드리뷰에서 고친 사항을 적어주세요. 재PR 시에만 사용해 주세요! (재PR 아닌 경우 삭제) -->

## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->

jhk01007 committed:
[QFEED-52] Chore: Spring ai 설정 추가

jhk01007 committed:
[QFEED-52] Rename: external api 디렉토리 구조변경

jhk01007 committed:
[QFEED-52] Chore: Redis Vector Database 의존성 추가

jhk01007 committed:
[QFEED-52] Chore: Redis Vector Database 관련 설정 정보 추가

jhk01007 committed:
[QFEED-52] Feat: Redis Vector Database 빈 수동 등록

jhk01007 committed:
[QFEED-52] Chore: Application 모듈에 External-api 모듈 의존성 추가

jhk01007 committed:
[QFEED-52] Feat: Redis Vector 기본 쿼리 로직 추가

- 텍스트 하나 저장
- 텍스트 여러 개 저장
- 비슷한 텍스트 찾기

jhk01007 committed:
[QFEED-52] Feat: QuestionRepository에 Question 저장 메서드 추가

jhk01007 committed:
[QFEED-52] Refactor: 생성된 질문 형식 내에 변수명 대문자로 변경

- CategoryName Enum과 동일하게 하였음

jhk01007 committed:
[QFEED-52] Feat: Category 엔티티에 @Getter 추가

jhk01007 committed:
[QFEED-52] Feat: SQL 쿼리 로그 추가

jhk01007 committed:
[QFEED-52] Feat: client 요청 오류 발생 시 자동으로 재시도하도록 설정

jhk01007 committed:
[QFEED-52] Feat: Category 정적 팩토리 메서드 추가

jhk01007 committed:
[QFEED-52] Feat: CategoryRepository 추가

jhk01007 committed:
[QFEED-52] Comment: RedisVectorStoreConfig 주석 수정

jhk01007 committed:
[QFEED-52] Feat: 질문 생성 기능 및 중복 검증 기능 추가

jhk01007 committed:
[QFEED-52] Feat: 토큰 계산 유틸 추가

jhk01007 committed:
[QFEED-52] Refactor: 질문 저장 시점을 모든 카테고리 질문들의 중복 검증이 끝났을 때로 변경

- 질문 생성 도중 오류 발생시 데이터 불일치를 방지.

jhk01007 committed:
[QFEED-52] Chore: ddl-auto create로 원상복구

jhk01007 committed:
[QFEED-52] Feat: CategoryRepository 저장 메서드 추가

LeeHyeonseob commented on module-external-api/src/main/java/com/wsws/moduleexternalapi/feed/config/RedisVectorStoreConfig.java:
RedisTemplate 대신에 Jedis를 선택하신 이유가 있나요?

jhk01007 replied:
우선 저는 저희 서비스내에 사용되는 Redis와는 별개의 인스턴스를 만들고자 했습니다.
그래서 별도의 빈을 만들어야 했는데 Redis Vector DB 라이브러리 버그인지는 모르겠지만 Spring Data Redis 라이브러리와 함께 사용시 빈이 만들어지지 않는 현상이 발생해 별도의 Redis 연결 라이브러리인 Jedis를 사용해 수동으로 빈을 구성하게 되었습니다.
참고로 공식문서에서도 수동으로 빈 구성시 Jedis를 사용하고 있습니다.
https://docs.spring.io/spring-ai/reference/api/vectordbs/redis.html#_manual_configuration

jhk01007 committed:
[QFEED-52] Refactor: 질문 블랙리스트 자료구조를 Map<String, List<String>> -> Map<String, Set<String>>로 변경

- 중복된 질문이 블랙리스트에 들어가 토큰이 증가하는 현상 해결

jhk01007 commented on module-application/src/main/java/com/wsws/moduleapplication/feed/service/scheduler/ScheduledQuestionCreateService.java:
```diff
@@ -0,0 +1,116 @@
+
+        // RDBMS에 질문 저장
+        for (String c : questionTempStore.keySet()) {
+            Category category = categoryRepository.findByCategoryName(CategoryName.valueOf(c)); // 카테고리에 ID를 받아옴.
```
1. 트랜잭션을 어떻게 걸지
2. 쿼리를 줄일 수 있으면 좋다.

jhk01007 replied:
인스턴스가 여러 개가 되면 ?

HDyoung commented on module-application/src/main/java/com/wsws/moduleapplication/feed/service/scheduler/ScheduledQuestionCreateService.java:
```diff
@@ -0,0 +1,116 @@
+    private final QuestionRepository questionRepository;
+    private final CategoryRepository categoryRepository;
+
+    private List<String> categories;
```
보니까 Category라는 도메인을 만들어두었던데 string이 아닌 도메인을 사용하면 더 명확할거 같네요

HDyoung commented on module-application/src/main/java/com/wsws/moduleapplication/feed/service/scheduler/ScheduledQuestionCreateService.java:
```diff
@@ -0,0 +1,116 @@
+    private final CategoryRepository categoryRepository;
+
+    private List<String> categories;
+    private Map<String, Set<String>> questionBlackListMap; // 카테고리 별로 중복된 질문을 담는 블랙 리스트
```
String이 아닌 CategoryId로 쓰면 context상 더 좋을거 같습니다

jhk01007 replied:
자료구조를 그대로 넘겨줌으로써 String을 통해 프롬프트를 작성하기 쉽게 하기위해 그런 선택을 했습니다.
해당 부분 수정에는 시간이 걸릴 것 같아 추후에 수정하도록하겠습니다! 감사합니다.

HDyoung commented on module-application/src/main/java/com/wsws/moduleapplication/feed/service/scheduler/ScheduledQuestionCreateService.java:
```diff
@@ -0,0 +1,116 @@
+        initList(); // 리스트 초기화
+
+        // 모든 카테고리의 질문들이 생성될 때 까지 반복
+        while (!categories.isEmpty()) {
```
while문을 보면 위험하다는 생각이듭니다. categories가 계속해서 지워지지 않으면 무한루프가 돌게 될텐데 만일의 상황을 대비한 탈출문구가 필요해 보여요

jhk01007 replied:
반영하겠습니다!

jhk01007 replied:
해당 부분은 어떻게 해야할지 고민을 조금 더 해봐야 할 것 같습니다.
우선 TODO로 처리하고 추후 구현하도록 하겠습니다!

HDyoung commented on module-application/src/main/java/com/wsws/moduleapplication/feed/service/scheduler/ScheduledQuestionCreateService.java:
```diff
@@ -0,0 +1,116 @@
+
+        // 모든 카테고리의 질문들이 생성될 때 까지 반복
+        while (!categories.isEmpty()) {
+            Map<String, String> createdQuestions = questionGenerateClient.createQuestions(categories, questionBlackListMap);
```
클래스 상단에 Transactional이 걸려 있는데 트랙젝션 오픈한채로 외부 api를 호출하는건 지양하는게 좋습니다. 외부 api호출이 오래걸리면 트랜젝션이 같이 열려 있는 비효율적인 구조가 됩니다

jhk01007 replied:
데이터 저장 부분을 별도의 클래스로 분리하고 그 클래스 내에서 @Transactional을 걸어주는식으로 수정하였습니다

HDyoung commented on module-application/src/main/java/com/wsws/moduleapplication/feed/service/scheduler/ScheduledQuestionCreateService.java:
```diff
@@ -0,0 +1,116 @@
+
+    private List<String> categories;
+    private Map<String, Set<String>> questionBlackListMap; // 카테고리 별로 중복된 질문을 담는 블랙 리스트
+    private Map<String, String> questionTempStore; // 카테고리 별로 검증이 끝난 질문 임시 저장소
```
지금은 단일 스레드로 도는걸라 영향이 없겠지만, 스케줄러를 여러 스레드에서 돌린다고 했을때는 ConcurrentMap같은 thread safe한 자료구조를 써야한다는걸 알고 있으면 좋을거 같아 말씀드립니다. 위에 List도 마찬가지에요

jhk01007 replied:
반영했습니다!

HDyoung commented on module-application/src/main/java/com/wsws/moduleapplication/feed/service/scheduler/ScheduledQuestionCreateService.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    private void addQuestionsToBlackListMap(String categoryName, String question, List<String> similarQuestions) {
+        if (questionBlackListMap.containsKey(categoryName)) { // 생성된 질문 블랙리스트에 저장
```
if else를 나눠쓰지 않고 map의 내장함수 computeIfAbsent를 활용해서 할 수도 있습니다.

jhk01007 replied:
반영했습니다!

jhk01007 committed:
[QFEED-52] Refactor: Category 조회 쿼리 6번 나가는 것을 1번에 다 가져오도록 변경

HDyoung commented on module-application/src/main/java/com/wsws/moduleapplication/feed/service/scheduler/ScheduledQuestionCreateService.java:
```diff
@@ -0,0 +1,116 @@
+     * 매일 23시 30분에 질문을 생성
+     */
+    @Scheduled(cron = "0 30 23 * * ?")
+    public void createQuestion() {
```
에러처리에 대한 로직이 있으면 좋을거 같네요

jhk01007 replied:
반영했습니다!

jhk01007 committed:
[QFEED-52] Refactor: 멀티스레드 환경을 고려한 자료구조로 변경

jhk01007 committed:
[QFEED-52] Refactor: Map의 computeIfAbsent 적용

jhk01007 committed:
[QFEED-52] Feat: 무한루프 방지를 위한 재시도 제한 기능 추가

jhk01007 committed:
[QFEED-52] Refactor: 데이터 접근하는 로직을 별도의 클래스로 분리하여 @Transactional 적용

jhk01007 committed:
[QFEED-52] Comment: TODO 주석 추가

jhk01007 committed:
[QFEED-52] Chore: External-api -> Domain 의존성 추가

jhk01007 committed:
[QFEED-52] Refactor: Application에서 Domain 계층만 의존하도록 변경

jhk01007 committed:
[QFEED-52] Refactor: ScheduledQuestionCreateService 내에 External-api 모듈 의존성 제거

- TokenCalculateUtil를 통해 External-api에 의존하던 상황 해결
- QuestionGenerateClient 인터페이스에 필요한 메서드 추가

jhk01007 committed:
[QFEED-52] Refactor: ScheduledQuestionCreateService에 AI 관련 의존성 제거

- ScheduledQuestionCreateService의 책임과 QuestionAIService 책임을 완전히 분리
